lib/rex/proto/ldap/server/auth.rb

```diff
@@ -0,0 +1,352 @@
+                  type: 'ntlm'
+                })
+                arg[:error_msg] = 'NTLM2_session Hash correspond to an empty password, ignoring ... '
+                return
+              end
+
+              hash = [
+                  lm_hash || '0' * 48,
+                  nt_hash || '0' * 48
+              ].join(':').gsub(/\n/, '\\n')
+              arg[:private] = hash
+            else
+                return
+            end
+            arg[:domain] = domain
+            arg[:user] = user
+            arg[:private_type] = :ntlm_hash
+            arg
+          end
+
+          def search_res(filter, msgid, attrflt = :all)
+            if @ldif.nil? || @ldif.empty?
+              attrs = []
+              if attrflt.is_a?(Array)
+                attrflt.each do |at|
+                  attrval = [Rex::Text.rand_text_alphanumeric(10)].map(&:to_ber).to_ber_set
+                  attrs << [at.to_ber, attrval].to_ber_sequence
+                end
+                dn = "dc=#{Rex::Text.rand_text_alphanumeric(10)},dc=#{Rex::Text.rand_text_alpha(4)}"
+                appseq = [
+                  dn.to_ber,
+                  attrs.to_ber_sequence
+                ].to_ber_appsequence(Net::LDAP::PDU::SearchReturnedData)
+                [msgid.to_ber, appseq].to_ber_sequence
+              end
+            else
+              ldif.map do |bind_dn, entry|
+                next unless filter.match(entry)
+        
+                attrs = []
+                entry.each do |k, v|
+                  if attrflt == :all || attrflt.include?(k.downcase)
+                    attrvals = v.map(&:to_ber).to_ber_set
+                    attrs << [k.to_ber, attrvals].to_ber_sequence
+                  end
+                end
+                appseq = [
+                  bind_dn.to_ber,
+                  attrs.to_ber_sequence
+                ].to_ber_appsequence(Net::LDAP::PDU::SearchReturnedData)
+                [msgid.to_ber, appseq].to_ber_sequence
+              end.compact.join
+            end
+          end
+        
+          def parse_type1_domain(message)
+            domain = nil
+            workstation = nil
+        
+            reqflags = message[12, 4]
+            reqflags = reqflags.unpack('V').first
+        
+            if (reqflags & NTLM_CONST::NEGOTIATE_DOMAIN) == NTLM_CONST::NEGOTIATE_DOMAIN
+              dom_len = message[16, 2].unpack('v')[0].to_i
+              dom_off = message[20, 2].unpack('v')[0].to_i
+              domain = message[dom_off, dom_len].to_s
+            end
+            if (reqflags & NTLM_CONST::NEGOTIATE_WORKSTATION) == NTLM_CONST::NEGOTIATE_WORKSTATION
+              wor_len = message[24, 2].unpack('v')[0].to_i
+              wor_off = message[28, 2].unpack('v')[0].to_i
+              workstation = message[wor_off, wor_len].to_s
+            end
+            [domain, workstation]
+          end
+
+          def encode_ldapsasl_response(msgid, code, dn, msg, creds, tag)
```

Comment:
Expose method as a utility helper since it does not require any state.

```suggestion
          def self.encode_ldapsasl_response(msgid, code, dn, msg, creds, tag)
```

Also is the format of this response consistent across all `SASL` responses? If so possibly rename this as `encode_sasl_response`.

Reply:
Based on the comment about `continued_sasl_auth` this might not need to be exposed at all, just made private and called in `continued_sasl_auth` or whatever that helper method ends up named.

Reply:
I thought that the encoding of `SASL` response can be handled within the already present `encode_ldap_response` using a case statement. Your thoughts?

Reply:
That might be reasonable depending on how consistent the `SASL` process is. Evaluating a response to determine if is for in progress `auth` may be a scoping the responsibilities of `encode_ldap_response` a bit broadly unless there are more conditional encoding patterns that need a similar path.